scipy/stats/_stats_py.py

```diff
@@ -4125,15 +4261,20 @@ def pearsonr(x, y):
     # If an input is constant, the correlation coefficient is not defined.
     if (x == x[0]).all() or (y == y[0]).all():
         warnings.warn(PearsonRConstantInputWarning())
-        return np.nan, np.nan
+        result = PearsonRResult(statistic=np.nan, pvalue=np.nan, n=n,
+                                alternative=alternative)
```

Comment:
One thing I am worried about is adding `_axis_nan_policy_factory` to these result tuples. First, how do we retrieve the `n` and `alternative` attributes when the decorator is applied? And if we manage to do that, `confidence_interval` is not vectorized to handle multidimensional inputs.

Do we want to add `_axis_nan_policy_factory` to `pearsonr`? I see it is listed in the [spreadsheet](https://docs.google.com/spreadsheets/d/1yBhu3Ihy9_xhDh5N9GgoNVyFbUHKWvt_lCuHm5BZ06I/edit#gid=0) but not sure if all the functions in the sheet need the decorator.

Reply:
gh-15632 shows how to get attributes out of these bunches. I need to merge `main` now that the `keepdims` PR has merged; gh-15632 will be simpler after that because it made several similar changes.

The confidence interval question is a really good one. I need to think about that. One quick answer is that I don't see the decorator as a forever-solution. Ideally, in the (perhaps very) long run, calculations will be natively vectorized, NumPy and SciPy functions will offer better support for omitting NaNs on their own, etc., so that maybe we can get the desired behavior without a decorator. But yeah, I don't like the idea of having to choose between an object with methods like `confidence_interval` and a vectorized result. It can certainly be made to work; but I'll need to think more about a clean way.

Reply:
I think this is pretty simple, actually. It's just a matter of updating our definition of the `PearsonRResult` object, which can easily be vectorized, and this is after all the hard stuff (omitting nans, tuple axis, etc.) has already been taken care of. Inside the decorator, `PearsonRResult` is the `tuple_to_result` object. It will be passed an array of statistics, `n`s, etc. The `confidence_interval` method just needs to be vectorized to act on _each element_, not along an axis, so it's not so bad.
Does that sound right?

Reply:
> an array of statistics, an array of `n`s, etc.

The decorator won't pass an array of `n`s and `alternative`s to `tuple_to_result` object since the result returned by `pearsonr` doesn't contain them as iterables. Even if it did, `alternative` is a string and the decorator treats each output as a float64. I don't know how would one immediately solve this problem.

> The `confidence_interval` method just needs to be vectorized to act on _each element_, not along an axis, so it's not so bad.

Yeah, this is correct.

Reply:
>  since the result returned by pearsonr doesn't contain them as iterables

Again, see gh-15632, which shows the little change that lets us do that. I'll need to think about `alternative` some more. But it's code. There will be a way.

Reply:
First thing that comes to mind that would work without change to the existing structure: `result_to_tuple` can turn `alternative` into a float corresponding with one of the three alternatives (e.g. -1 for less, 0 for two-sided, 1 for greater). `tuple_to_result` can look at any element of the resulting array (they will all be the same) and use that to set the `alternative` field of the `PearsonRResult` object to the appropriate string. (Since they will all be the same, no sense making it an array of strings. But `n` could, in general, vary if there are masked elements or omitted NaNs).

Reply:
> First thing that comes to mind that would work without change to the existing structure: `result_to_tuple` can turn `alternative` into a float corresponding with one of the three alternatives (e.g. -1 for less, 0 for two-sided, 1 for greater). `tuple_to_result` can look at any element of the resulting array (they will all be the same) and use that to set the `alternative` field of the `PearsonRResult` object to the appropriate string.

That could work! One question: IIUC, the decorator could return an array of `nan` as result (i.e. `alternative=nan` could be passed to `tuple_to_result`). In that case, how does one decode the value of `alternative`? (I guess it doesn't matter since the result is just going to be nan anyways.)

Reply:
Yeah I don't think it matters. 
And I intended for us to allow that behavior - returning an all-nan array when nan_policy='propagage' - to be overridden. When `@_axis_nan_policy(....)` is called, we could pass a callable, 
But we don't really need to.